Certification Table Endpoints done

AmanDeol7 opened the pull request:
## Description of Changes

### 1. Certificate Management API (CRUD Endpoints)
- **Added full CRUD endpoints** for certificates in `certificate_controller.py`:
  - **Create, List, Get, Update, and Delete** certification records via REST API routes.
  - Added pagination, sorting, and filtering support in the list endpoint.
  - Integrated FastAPI dependencies for session management and type validation.

### 2. Data Transfer Objects (DTOs)
- **Created `certification_entity.py`**:
  - Introduced `CreateCertification`, `UpdateCertification`, and `ReadCertification` Pydantic models with validation for required fields.
  - Models support all relevant certificate attributes, including tools and organization info.

### 3. Repository Layer
- **Added `certifications_repository.py`**:
  - Implements methods for creating, listing, retrieving, updating, and deleting certifications in the database.
  - Supports filtering by issuing organization and user, as well as sorting and pagination.

### 4. Service Layer Update
- **Refactored `certificate_service.py`**:
  - Added `CertificateGeneratorService` class to handle business logic for certification CRUD operations.
  - Handles error scenarios such as certificate not found.

### 5. Error Handling Enhancements
- **Extended exception handling in `error_handlers.py`**:
  - Custom handlers for `CertificateNotFound` and `CertificationsUnAvailable` exceptions.
  - Returns detailed API error responses for certificate-specific errors.

### 6. Custom Exceptions
- **Updated `user_exceptions.py`**:
  - Added `CertificateNotFound`, `CertificateAlreadyExists`, and `CertificationsUnAvailable` exception classes.

### 7. Error Codes
- **Expanded `error_codes.py`**:
  - Added specific error codes for certificate-related operations:
    - Database errors, validation failures, not found, and permission issues.

---

**In summary:**  
This commit introduces a complete certificate management system to the backend, including API endpoints, validation, database integration, and robust error handling for user certifications.

JRS296 reviewed (changes requested):
@AmanDeol7 The Certificate Controller has to do with the Dijkstra Certificates. The Endpoints are to deal with Certifications (Plural). They are two seperate controller cases. The distinction was probably not clear enough. Kindly revert changes to the certificate controller, rename it to dijkstra_certificate, and then create a new controller class called certifications_controller (keep in mind certifications, not certificate (singular))

Feel free to DM me for any further queries

AmanDeol7 commented on app/Controllers/User/certificate_controller.py:
renamed it to certifications_controller.py and created a new file called dijkstra_certificate controller with the previous code

AmanDeol7 commented:
## Description of Changes in latest commit

### 1. Refactored Certificate Logic to Certifications

- **Renamed Controllers and Services:**
  - The original `certificate_controller.py` and `certificate_service.py` were renamed to `certifications_controller.py` and `certifications_service.py` respectively.
  - All references to "Certificate" were changed to "Certification" throughout the controllers, services, and exception handling logic for consistency.

- **Modified API Routing:**
  - The certificate-related API routes now use `/Dijkstra/v1/certifications` and are tagged as `Certifications` instead of `Certificate`.

- **Exception Handling Updates:**
  - Custom exception names and their handlers were updated, e.g., `CertificateNotFound` → `CertificationNotFound`, `CertificationsUnAvailable` → `CertificationUnAvailable`, and so forth.
  - Exception handler registration and raised errors were refactored to align with the new naming.

- **Error Code Updates:**
  - Error code constants in `error_codes.py` were changed from `USER_CERTIFICATE_*` to `USER_CERTIFICATION_*` for database, server, authentication, validation, and not found errors.

### 2. Split Download Certificate Functionality

- The logic for certificate download (`/download/{userName}`) and its related health check was moved out of the certifications controller into a dedicated `dijkstra_certificate_controller.py` and `dijkstra_certificate_service.py`.
- This separation clarifies the distinction between generic certificate generation (download) and certification CRUD operations.

### 3. Main Application Routing Changes

- In `main.py`, the inclusion of routers was updated:
  - Removed the old `certificate_controller`.
  - Added the new `dijkstra_certificate_controller` (for download endpoints) and `certifications_controller` (for CRUD and listing certifications).
- Ensured that both download and certification endpoints are registered and available.

### 4. General Cleanup and Consistency

- Standardized naming conventions for classes, methods, and error handling.
- Removed redundant code and fixed minor formatting issues.

---

**In summary:**  
This commit refactors the backend’s certificate management to use more precise "certification" terminology, splits certificate generation from certification CRUD, updates API routing, error handling, and cleans up code for clarity and maintainability.

JRS296 commented on app/Controllers/User/certifications_controller.py:
```diff
@@ -0,0 +1,66 @@
+
+router = APIRouter(prefix="/Dijkstra/v1/certifications", tags=["Certifications"])
+
+@router.get('/health', status_code=200)
+async def root():
+    logger.info("Health Endpoint Triggered")
+    return {"status": 200, 'message': 'Certifications Health Endpoint Triggered!!!'}
```
This can be removed

JRS296 commented on app/Controllers/error_handlers.py:
```diff
@@ -110,6 +110,26 @@ async def work_experience_not_found_handler(request: Request, exc: WorkExperienc
             status=404
         )
 
+    @app.exception_handler(CertificationNotFound)
+    async def certification_not_found_handler(request: Request, exc: CertificationNotFound):
+        logger.warning(f"Certificate not found: {exc.certificate_id}")
+        raise_api_error(
+            code=ErrorCodes.USER_CERTIFICATE_NF_A01,
+            error="Certificate not found",
+            detail=str(exc),
+            status=404
+        )
+
+    @app.exception_handler(CertificationUnAvailable)
+    async def certifications_unavailable_handler(request: Request, exc: CertificationUnAvailable):
+        logger.info("No certifications available")
+        raise_api_error(
+            code=ErrorCodes.USER_CERTIFICATE_NF_A01,
+            error="No certifications found",
+            detail=str(exc),
+            status=404
+        )
+
```
What is the difference between these two? They seem to be the same

AmanDeol7 replied:
Certificate Not found is used for GET request of a particular certificate, certifications unavailable is used when I do a list-all GET query with certain filters and nothing shows up

JRS296 replied:
There's little disctinction between the two, so I would recommend that you remove CertificationUnAvailable

JRS296 reviewed (changes requested):
Kindly make the requested changes

AmanDeol7 commented:
> May I know what these SS are for?
> 
> 

clarification on the error codes
